packages/babel-plugin-proposal-do-expressions/src/index.ts

```diff
@@ -10,19 +12,239 @@ export default declare(api => {
+        t.blockStatement([t.returnStatement(body.node as t.Expression)]),
+      );
+      flattenStatement(newBody.get("body")[0]);
+      break;
+    }
+    path = path.parentPath;
+  } while (path);
+}
+
+function flattenStatement(path: NodePath<t.Statement>) {
+  const effects: t.Statement[] = [];
+  switch (path.type) {
+    case "WhileStatement": {
+      const test = path.get("test");
+      pushEffect(test);
+      path.set("test", t.booleanLiteral(true));
+      path.set(
+        "body",
+        t.blockStatement([
+          ...effects,
+          t.ifStatement(t.unaryExpression("!", test.node), t.breakStatement()),
+          path.node.body,
+        ]),
+      );
+      break;
+    }
+    case "DoWhileStatement": {
+      const test = path.get("test");
+      pushEffect(test);
+      path.set("test", t.booleanLiteral(true));
+      path.set(
+        "body",
+        t.blockStatement([
+          path.node.body,
+          ...effects,
+          t.ifStatement(t.unaryExpression("!", test.node), t.breakStatement()),
+        ]),
+      );
+      break;
+    }
+    case "ExpressionStatement":
+      pushEffect(path.get("expression"), false);
+      path.replaceWithMultiple(effects);
+      break;
+    default:
+      flattenNormal(path);
+      path.replaceWithMultiple([...effects, path.node]);
+  }
+
+  function pushEffect(path: NodePath<t.Expression>, needResult = true) {
+    const { node } = path;
+    if (node.type === "DoExpression") {
+      effects.push(node.body);
+      if (needResult) {
+        const uid = path.get("body").scope.generateDeclaredUidIdentifier("do");
+        captureDoExpressionResult(node.body, uid);
+        path.replaceWith(uid);
+      }
+    } else {
+      if (isTopLevelSideEffectFree(node)) {
+        flattenNormal(path);
+      } else {
+        if (traverse.hasType(node, "DoExpression")) {
+          flattenNormal(path);
+        }
+        if (needResult) {
+          const uid = path.scope.generateDeclaredUidIdentifier("do");
+          effects.push(
+            t.expressionStatement(
+              t.assignmentExpression("=", t.cloneNode(uid), node),
+            ),
+          );
+          path.replaceWith(uid);
+        } else {
+          effects.push(t.expressionStatement(node));
+        }
+      }
+    }
+  }
+
+  function flattenNormal(path: NodePath) {
+    const expressions: NodePath<t.Expression>[] = [];
+    path.traverse({
+      Statement(path) {
+        path.skip();
+      },
+      Expression(path) {
+        expressions.push(path);
+        path.skip();
+      },
+    });
+    for (let i = expressions.length - 1; i >= 0; i--) {
+      const path = expressions[i];
+      if (
+        path.node.type !== "DoExpression" &&
+        !traverse.hasType(path.node, "DoExpression")
+      ) {
+        expressions.pop();
+      } else {
+        break;
+      }
+    }
+    expressions.forEach(path => pushEffect(path));
+  }
+}
+
+function captureDoExpressionResult(
```

Comment:
It seems to me this utility shares a similar purpose with the `NodePath#getCompletionRecords`, can we apply the capture transform to the return result of `NodePath#getCompletionRecords`?